File: Dev/C++/Mascaret/src/VEHA/Entity/Entity.cpp
#include "VEHA/Entity/Entity.h"
#include "VEHA/Entity/ShapeSpecification.h"
#include "VEHA/Entity/AnimationSpecification.h"
#include "VEHA/Entity/SoundSpecification.h"
#include "VEHA/Entity/PointSpecification.h"
#include "VEHA/Entity/PathSpecification.h"
#include "VEHA/Entity/RotationVector.h"
#include "VEHA/Entity/Vector3.h"
#include "VEHA/Entity/EntityClass.h"
#include "VEHA/Entity/VirtualRealityComponentFactory.h"
#include "VEHA/Kernel/CallbackManager.h"
#include "MascaretApplication.h"

namespace VEHA
{

Entity::Entity(const string& name, shared_ptr<EntityClass> cl) :
    InstanceSpecification(name, cl),
    _activeShape(),
    _referentialPoint(MascaretApplication::getInstance()->getVRComponentFactory()->createPoint()),
    _parent()
{
}

Entity::~Entity()
{
}

bool Entity::hasChild(shared_ptr<Entity> entity)
{
	bool hasChild = false;
	for (unsigned int i = 0; i < _children.size(); i++)
	{
		if (entity == _children[i]) 
		{
			hasChild = true;
			break;
		}
	}
	if (hasChild) return hasChild;
	else
	{
		int i = 0;
		while (i < _children.size() && !hasChild)
			hasChild = _children[i++]->hasChild(entity); 
	}
	return hasChild;
}

shared_ptr<Entity> Entity::getChild(string name)
{
  for (size_t i = _children.size(); i--; )
    if(_children[i]->getName() == name)
      return _children[i];

  return shared_ptr<Entity>();
}

void Entity::setActiveShape(const string& name)
{
  if (getClassifier()->hasAttribute(name)) // TODO: copie des proprietes dans la nouvelle shape
  {
    shared_ptr<Slot> s = getProperty(name);
    _activeShape = shared_dynamic_cast<ShapeSpecification>(s->getValue());
    if (_activeShape)
      _activeShape->setEntity(shared_dynamic_cast<Entity>(shared_from_this()));
    else
      cerr << getName() << "->setActiveShape(" << name << ") : the property is undefined" << endl;
  }
}

void Entity::setLocalPosition(const Vector3& pos)
{
  CallbackManager_changesMethod::addChangeMethodSignal(&_cbChangePosition); //TODO pb si le slot disparait
  _referentialPoint->setLocalPosition(pos);
  fireChangePosition();
}

void Entity::translate(const Vector3& pos)
{
  CallbackManager_changesMethod::addChangeMethodSignal(&_cbChangePosition); //TODO pb si le slot disparait
  _referentialPoint->translate(pos);
  fireChangePosition();
}

Vector3 Entity::getLocalPosition()
{
  return _referentialPoint->getLocalPosition();
}

void Entity::setGlobalPosition(const Vector3& pos)
{
  CallbackManager_changesMethod::addChangeMethodSignal(&_cbChangePosition); //TODO pb si le slot disparait
  _referentialPoint->setGlobalPosition(pos);
  fireChangePosition();
}
Vector3 Entity::getGlobalPosition()
{
  return _referentialPoint->getGlobalPosition();
}
void Entity::setLocalRotation(const RotationVector& rot)
{
  _referentialPoint->setLocalRotation(rot);
  fireChangePosition();
}

RotationVector Entity::getLocalRotation()
{
  CallbackManager_changesMethod::addChangeMethodSignal(&_cbChangePosition); //TODO pb si le slot disparait
  return _referentialPoint->getLocalRotation();
}
void Entity::setGlobalRotation(const RotationVector& rot)
{
  _referentialPoint->setGlobalRotation(rot);
  fireChangePosition();
}
RotationVector Entity::getGlobalRotation()
{
  CallbackManager_changesMethod::addChangeMethodSignal(&_cbChangePosition); //TODO pb si le slot disparait
  return _referentialPoint->getGlobalRotation();
}

void Entity::rotate(const RotationVector& ori)
{
  CallbackManager_changesMethod::addChangeMethodSignal(&_cbChangePosition); //TODO pb si le slot disparait
  _referentialPoint->rotate(ori);
  fireChangePosition();
}

void Entity::rotate(const Vector3& ori)
{
  CallbackManager_changesMethod::addChangeMethodSignal(&_cbChangePosition); //TODO pb si le slot disparait
  _referentialPoint->rotate(ori);
  fireChangePosition();
}

void Entity::setLocalOrientation(const Vector3& ori)
{
  Vector3 ori2 = ori;
  _referentialPoint->setLocalOrientation(ori2);
  fireChangePosition();
}

Vector3 Entity::getLocalOrientation()
{
  CallbackManager_changesMethod::addChangeMethodSignal(&_cbChangePosition); //TODO pb si le slot disparait
  return _referentialPoint->getLocalOrientation();
}
void Entity::setGlobalOrientation(const Vector3& ori)
{
  Vector3 ori2 = ori;
  _referentialPoint->setGlobalOrientation(ori2);
  fireChangePosition();
}
Vector3 Entity::getGlobalOrientation()
{
  CallbackManager_changesMethod::addChangeMethodSignal(&_cbChangePosition); //TODO pb si le slot disparait
  return _referentialPoint->getGlobalOrientation();
}
void Entity::setVisibility(bool visible)
{
  getActiveShape()->setVisibility(visible);
}

bool Entity::getVisibility()
{
  return getActiveShape()->getVisibility();
}
vector<shared_ptr<ShapeSpecification> > Entity::getShapes()
{
  vector<shared_ptr<ShapeSpecification> > vect;
  const map<string, shared_ptr<Slot> >& slots = getSlots();
  for (map<string, shared_ptr<Slot> >::const_iterator i = slots.begin(); i != slots.end(); i++)
  {
    const map<string, shared_ptr<ValueSpecification> >& values = i->second->getValues();
    if (values.size() == 1)
    {
      for (map<string, shared_ptr<ValueSpecification> >::const_iterator i = values.begin();
          i != values.end(); i++)
      {
        shared_ptr<ShapeSpecification> sh = shared_dynamic_cast<ShapeSpecification>(i->second);
        if (sh) vect.push_back(sh);
      }
    }
  }
  return vect;
}

shared_ptr<ShapeSpecification> Entity::getShape(const string& name)
{
  if (getClassifier()->hasAttribute(name))
  {
    shared_ptr<Slot> prop = getProperty(name);
    if (prop) return shared_dynamic_cast<ShapeSpecification>(prop->getValue());
  }
  return shared_ptr<ShapeSpecification>();
}

vector<shared_ptr<AnimationSpecification> > Entity::getAnimations()
{
  vector<shared_ptr<AnimationSpecification> > vect;
  const map<string, shared_ptr<Slot> >& slots = getSlots();
  for (map<string, shared_ptr<Slot> >::const_iterator i = slots.begin(); i != slots.end(); i++)
  {
    const map<string, shared_ptr<ValueSpecification> >& values = i->second->getValues();
    if (values.size() == 1)
    {
      for (map<string, shared_ptr<ValueSpecification> >::const_iterator i = values.begin();
          i != values.end(); i++)
      {
        shared_ptr<AnimationSpecification> sh = shared_dynamic_cast<AnimationSpecification>(
            i->second);
        if (sh) vect.push_back(sh);
      }
    }
  }
  return vect;
}
shared_ptr<AnimationSpecification> Entity::getAnimation(const string& name)
{
  if (getClassifier()->hasAttribute(name))
  {
    shared_ptr<Slot> prop = getProperty(name);
    if (prop) return shared_dynamic_cast<AnimationSpecification>(prop->getValue());
  }
  return shared_ptr<AnimationSpecification>();
}

vector<shared_ptr<SoundSpecification> > Entity::getSounds()
{
  vector<shared_ptr<SoundSpecification> > vect;
  const map<string, shared_ptr<Slot> >& slots = getSlots();
  for (map<string, shared_ptr<Slot> >::const_iterator i = slots.begin(); i != slots.end(); i++)
  {
    const map<string, shared_ptr<ValueSpecification> >& values = i->second->getValues();
    if (values.size() == 1)
    {
      for (map<string, shared_ptr<ValueSpecification> >::const_iterator i = values.begin();
          i != values.end(); i++)
      {
        shared_ptr<SoundSpecification> sh = shared_dynamic_cast<SoundSpecification>(i->second);
        if (sh) vect.push_back(sh);
      }
    }
  }
  return vect;
}
shared_ptr<SoundSpecification> Entity::getSound(const string& name)
{
  if (getClassifier()->hasAttribute(name))
  {
    shared_ptr<Slot> prop = getProperty(name);
    if (prop) return shared_dynamic_cast<SoundSpecification>(prop->getValue());
  }
  return shared_ptr<SoundSpecification>();
}

vector<shared_ptr<TopologicalSpecification> > Entity::getTopologicals()
{
  vector<shared_ptr<TopologicalSpecification> > vect;
  const map<string, shared_ptr<Slot> >& slots = getSlots();
  for (map<string, shared_ptr<Slot> >::const_iterator i = slots.begin(); i != slots.end(); i++)
  {
    const map<string, shared_ptr<ValueSpecification> >& values = i->second->getValues();
    if (values.size() == 1)
    {
      for (map<string, shared_ptr<ValueSpecification> >::const_iterator i = values.begin();
          i != values.end(); i++)
      {
        shared_ptr<TopologicalSpecification> sh = shared_dynamic_cast<TopologicalSpecification>(
            i->second);
        if (sh) vect.push_back(sh);
      }
    }
  }
  return vect;
}
shared_ptr<TopologicalSpecification> Entity::getTopological(const string& name)
{
  if (getClassifier()->hasAttribute(name))
  {
    shared_ptr<Slot> prop = getProperty(name);
    if (prop) return shared_dynamic_cast<TopologicalSpecification>(prop->getValue());
  }
  return shared_ptr<TopologicalSpecification>();
}

vector<shared_ptr<PointSpecification> > Entity::getPoints()
{
  vector<shared_ptr<PointSpecification> > vect;
  const map<string, shared_ptr<Slot> >& slots = getSlots();
  for (map<string, shared_ptr<Slot> >::const_iterator i = slots.begin(); i != slots.end(); i++)
  {
    const map<string, shared_ptr<ValueSpecification> >& values = i->second->getValues();
    if (values.size() == 1)
    {
      for (map<string, shared_ptr<ValueSpecification> >::const_iterator i = values.begin();
          i != values.end(); i++)
      {
        shared_ptr<PointSpecification> sh = shared_dynamic_cast<PointSpecification>(i->second);
        if (sh) vect.push_back(sh);
      }
    }
  }
  return vect;
}
shared_ptr<PointSpecification> Entity::getPoint(const string& name)
{
  if (getClassifier()->hasAttribute(name))
  {
    shared_ptr<Slot> prop = getProperty(name);
    if (prop) return shared_dynamic_cast<PointSpecification>(prop->getValue());
  }
  return shared_ptr<PointSpecification>();
}

vector<shared_ptr<PathSpecification> > Entity::getPaths()
{
  vector<shared_ptr<PathSpecification> > vect;
  const map<string, shared_ptr<Slot> >& slots = getSlots();
  for (map<string, shared_ptr<Slot> >::const_iterator i = slots.begin(); i != slots.end(); i++)
  {
    const map<string, shared_ptr<ValueSpecification> >& values = i->second->getValues();
    if (values.size() == 1)
    {
      for (map<string, shared_ptr<ValueSpecification> >::const_iterator i = values.begin();
          i != values.end(); i++)
      {
        shared_ptr<PathSpecification> sh = shared_dynamic_cast<PathSpecification>(i->second);
        if (sh) vect.push_back(sh);
      }
    }
  }
  return vect;
}
shared_ptr<PathSpecification> Entity::getPath(const string& name)
{
  if (getClassifier()->hasAttribute(name))
  {
    shared_ptr<Slot> prop = getProperty(name);
    if (prop) return shared_dynamic_cast<PathSpecification>(prop->getValue());
  }
  return shared_ptr<PathSpecification>();
}

bool Entity::playAnimation(const string& animationName, int sens, double animationSpeed, bool cycle)
{
  shared_ptr<AnimationSpecification> anim = getAnimation(animationName);
  if (!anim) return false;
  bool result = true;
  anim->setShape(getActiveShape());
  result = anim->play(animationSpeed, sens, cycle);
  for (size_t i = 0; result && i < _children.size(); i++)
    result = _children[i]->playAnimation(animationName, sens, animationSpeed, cycle);
  return result;
}
bool Entity::stopAnimation(const string& animationName)
{
  shared_ptr<AnimationSpecification> anim = getAnimation(animationName);
  if (anim) return anim->stop();
  return false;
}
bool Entity::stopAllAnimations()
{
  vector<shared_ptr<AnimationSpecification> > anims = getAnimations();
  for (size_t i = 0; i < anims.size(); i++)
  {
    anims[i]->stop();
  }
  return true;
}
bool Entity::playSound(const string& name, double speed, bool cycle)
{
  shared_ptr<SoundSpecification> sound = getSound(name);
  if (!sound) return false;
  bool result = true;
  if (getActiveShape()) sound->attachTo(_referentialPoint);
  result = sound->play(speed, cycle);
  return result;
}

bool Entity::stopSound(const string& soundName)
{
  shared_ptr<SoundSpecification> sound = getSound(soundName);
  if (sound) return sound->stop();
  return false;
}
shared_ptr<PointSpecification> Entity::getViewPoint(const string& name)
{
  map<string, shared_ptr<PointSpecification> >::iterator it = _viewPoints.find(name);
  cerr << "[Entity Info] " << getName() << " has " << _viewPoints.size() << " viewpoints" << endl;
  if (it != _viewPoints.end()) return it->second;
  for (size_t i = 0; i < _children.size(); i++)
  {
    shared_ptr<PointSpecification> vp = _children[i]->getViewPoint(name);
    if (vp) return vp;
  }
  return shared_ptr<PointSpecification>();
}
void Entity::setParent(shared_ptr<Entity> e)
{
  if (_parent)
  {
    _parent->_removeChildren(shared_dynamic_cast<Entity>(this->shared_from_this()));

  }
  _referentialPoint->setParent(shared_ptr<PointSpecification>());
  _parent = e;
  if (_parent)
  {
    _referentialPoint->setParent(_parent->_referentialPoint);
    _parent->_addChildren(shared_dynamic_cast<Entity>(this->shared_from_this()));

  }
}

void Entity::addCallbackOnChangePosition( void (*f)(shared_ptr<Entity>) )
{
	_CallbackOnChangePosition.push_back(f);
}

void Entity::removeCallbackOnChangePosition( void (*f)(shared_ptr<Entity>) )
{
	unsigned int i;
	
	for (i = 0; i < _CallbackOnChangePosition.size(); i ++)
	{
		if (_CallbackOnChangePosition[i] == f) break;	
	}

	if (i < _CallbackOnChangePosition.size())
		_CallbackOnChangePosition.erase(_CallbackOnChangePosition.begin() + i);

}

void Entity::fireChangePosition(void)
{
	for (unsigned i = 0; i < _CallbackOnChangePosition.size(); i ++)
	{
		(*_CallbackOnChangePosition[i])(shared_dynamic_cast<Entity>(this->shared_from_this()));
	}
}



} /// End: namespace VEHA

File: Dev/CS/Mascaret/Mascaret/VEHA/Entity/Entity.cs
using System;
using System.Collections.Generic;

namespace Mascaret
{
    public class Entity : InstanceSpecification
    {

        private ShapeSpecification activeShape = null;
        public ShapeSpecification ActiveShape
        {
            get { return activeShape; }
            set { activeShape = value; }
        }


        private PointSpecification referentialPoint = null;
        public PointSpecification ReferentialPoint
        {
            get { return referentialPoint; }
            //set{referentialPoint=value;}
        }

        protected void setReferentialPoint(PointSpecification point)
        {
            referentialPoint = point;
        }


        private Dictionary<String, PointSpecification> viewPoints;
        public Dictionary<String, PointSpecification> ViewPoints
        {
            get { return viewPoints; }
            set { viewPoints = value; }
        }


        private Entity parent = null;
        public Entity Parent
        {
            get { return parent; }
            set { parent = value; }
        }


        private List<Entity> children = new List<Entity>();
        public List<Entity> Children
        {
            get { return children; }
            set { children = value; }
        }

        //missing callback

        public Entity(string name, EntityClass classe)
            : base(name, classe)
        {
            viewPoints = new Dictionary<string, PointSpecification>();
            //	referentialPoint= new  (UnityPointSpecification)ScriptableObject.CreateInstance("UnityPointSpecification");
            //	((UnityPointSpecification)referentialPoint).instantiate();//getGlobalPosition(),getGlobalRotation());

        }
        public Entity()
            : base("", null)
        {
            viewPoints = new Dictionary<string, PointSpecification>();
            //	referentialPoint= new  (UnityPointSpecification)ScriptableObject.CreateInstance("UnityPointSpecification");
            //	((UnityPointSpecification)referentialPoint).instantiate();//getGlobalPosition(),getGlobalRotation());

        }

        //Missing callback operations

        public void addChild(Entity element)
        {
            if (!children.Contains(element))
            {
                element.parent = this;
                children.Add(element);
            }
        }

        public bool hasChild(Entity element)
        {
            foreach (Entity child in children)
            {
                if (child == element)
                    return true;
            }

            foreach (Entity child in children)
            {
                return child.hasChild(element);
            }

            return false;
        }

        public Entity getChild(string name)
        {
            foreach (Entity child in children)
            {
                if (child.name.CompareTo(name) == 0)
                    return child;
            }
            return null;
        }

        public void setActiveShape(string name)
        {
            if (Classifier.hasAttribute(name))
            {
                Slot s = getProperty(name);
                if (s != null)
                {
                    activeShape = (ShapeSpecification)s.getValue();
                    if (activeShape != null)
                        activeShape.setEntity(this);
                    else
                        System.Console.WriteLine(this.name + ".setActiveShape(" + name + ") : the property is undefined");

                }
            }
        }

        public void setVisibility(bool v)
        {
            activeShape.setVisibility(v);
        }

        public bool getVisibility()
        {
            return activeShape.getVisibility();
        }

        public void translate(Vector3 pos)
        {
            //callback operation
            referentialPoint.translate(pos);
        }

        public void setLocalPosition(Vector3 pos)
        {
            //callback operation
            if (referentialPoint != null)
                referentialPoint.setLocalPosition(pos);
        }

        public Vector3 getLocalPosition()
        {
            return referentialPoint.getLocalPosition();
        }

        public void setGlobalPosition(Vector3 pos)
        {
            //callback operation
            referentialPoint.setGlobalPosition(pos);
        }

        public Vector3 getGlobalPosition()
        {
            return referentialPoint.getGlobalPosition();
        }

        public void setLocalRotation(RotationVector orientation)
        {
            //callback operation
          //  referentialPoint.setLocalRotation(orientation);
        }

        public RotationVector getLocalRotation()
        {
            return referentialPoint.getLocalRotation();
        }

        public void setGlobalRotation(RotationVector orientation)
        {
            //callback operation
            referentialPoint.setGlobalRotation(orientation);
        }

        public RotationVector getGlobalRotation()
        {
            return referentialPoint.getGlobalRotation();
        }

        public void setLocalOrientation(Vector3 pos)
        {
            //callback operation
            if (referentialPoint != null)
                referentialPoint.setLocalOrientation(pos);
        }

        public Vector3 getLocalOrientation()
        {
            //return referentialPoint.getLocalOrientation();
            return null;
        }

        public void setGlobalOrientation(Vector3 pos)
        {
            //callback operation
            referentialPoint.setGlobalOrientation(pos);
        }

        public Vector3 getGlobalOrientation()
        {
            return referentialPoint.getGlobalOrientation();
        }

        public List<ShapeSpecification> getShapes()
        {
            List<ShapeSpecification> vect = new List<ShapeSpecification>();
            foreach (Slot slot in Slots.Values)
            {
                Dictionary<string, ValueSpecification> values = slot.Values;
                if (values.Count == 1) //etrange ca aussi comme condition
                {
                    foreach (ValueSpecification valueSpec in values.Values)
                    {
                        if (valueSpec as ShapeSpecification != null)
                            vect.Add((ShapeSpecification)valueSpec);
                    }
                }
            }
            return vect;

        }

        public ShapeSpecification getShape(string name)
        {
            if (Classifier.hasAttribute(name))
            {
                Slot prop = getProperty(name);
                if (prop != null)
                    return (ShapeSpecification)prop.getValue();
            }

            return null;
        }

        public List<AnimationSpecification> getAnimations()
        {
            List<AnimationSpecification> vect = new List<AnimationSpecification>();
            foreach (Slot slot in Slots.Values)
            {
                Dictionary<string, ValueSpecification> values = slot.Values;
                if (values.Count == 1) //etrange ca aussi comme condition
                {
                    foreach (ValueSpecification valueSpec in values.Values)
                    {
                        if (valueSpec as AnimationSpecification != null)
                            vect.Add((AnimationSpecification)valueSpec);
                    }
                }
            }
            return vect;
        }

        public AnimationSpecification getAnimation(string name)
        {
            if (Classifier.hasAttribute(name))
            {
                Slot prop = getProperty(name);
                if (prop != null)
                    return (AnimationSpecification)prop.getValue();
            }
            return null;
        }

        public List<SoundSpecification> getSounds()
        {
            List<SoundSpecification> vect = new List<SoundSpecification>();
            foreach (Slot slot in Slots.Values)
            {
                Dictionary<string, ValueSpecification> values = slot.Values;
                if (values.Count == 1) //etrange ca aussi comme condition
                {
                    foreach (ValueSpecification valueSpec in values.Values)
                    {
                        if (valueSpec as SoundSpecification != null)
                            vect.Add((SoundSpecification)valueSpec);
                    }
                }
            }
            return vect;
        }

        public SoundSpecification getSound(string name)
        {
            if (Classifier.hasAttribute(name))
            {
                Slot prop = getProperty(name);
                if (prop != null)
                    return (SoundSpecification)prop.getValue();
            }
            return null;
        }

        public List<TopologicalSpecification> getTopologicals()
        {
            List<TopologicalSpecification> vect = new List<TopologicalSpecification>();
            foreach (Slot slot in Slots.Values)
            {
                Dictionary<string, ValueSpecification> values = slot.Values;
                if (values.Count == 1) //etrange ca aussi comme condition
                {
                    foreach (ValueSpecification valueSpec in values.Values)
                    {
                        if (valueSpec as TopologicalSpecification != null)
                            vect.Add((TopologicalSpecification)valueSpec);
                    }
                }
            }
            return vect;
        }

        public TopologicalSpecification getTopological(string name)
        {
            if (Classifier.hasAttribute(name))
            {
                Slot prop = getProperty(name);
                if (prop != null)
                    return (TopologicalSpecification)prop.getValue();
            }
            return null;
        }

        public List<PointSpecification> getPoints()
        {
            List<PointSpecification> vect = new List<PointSpecification>();
            foreach (Slot slot in Slots.Values)
            {
                Dictionary<string, ValueSpecification> values = slot.Values;
                if (values.Count == 1) //etrange ca aussi comme condition
                {
                    foreach (ValueSpecification valueSpec in values.Values)
                    {
                        if (valueSpec as PointSpecification != null)
                            vect.Add((PointSpecification)valueSpec);
                    }
                }
            }
            return vect;
        }

        public PointSpecification getPoint(string name)
        {
            if (Classifier.hasAttribute(name))
            {
                Slot prop = getProperty(name);
                if (prop != null)
                    return (PointSpecification)prop.getValue();
            }
            return null;
        }

        public List<PathSpecification> getPaths()
        {
            List<PathSpecification> vect = new List<PathSpecification>();
            foreach (Slot slot in Slots.Values)
            {
                Dictionary<string, ValueSpecification> values = slot.Values;
                if (values.Count == 1) //etrange ca aussi comme condition
                {
                    foreach (ValueSpecification valueSpec in values.Values)
                    {
                        if (valueSpec as PathSpecification != null)
                            vect.Add((PathSpecification)valueSpec);
                    }
                }
            }
            return vect;
        }

        public PathSpecification getPath(string name)
        {
            if (Classifier.hasAttribute(name))
            {
                Slot prop = getProperty(name);
                if (prop != null)
                    return (PathSpecification)prop.getValue();
            }
            return null;
        }

        //Default parameters sens = 1, animationSpeed = 1.0, cycle = true
        public bool playAnimation(string animationName, int sens, double animationSpeed, bool cycle)
        {
            AnimationSpecification anim = getAnimation(animationName);
            if (anim != null)
            {
                anim.setShape(activeShape);
                return anim.play(animationSpeed, sens, cycle);
            }
            else
                return false;
        }

        public bool playSound(string name, double speed, bool cycle)
        {
            SoundSpecification sound = getSound(name);
            if (sound != null)
            {
                //	sound.attachTo(referentialPoint);
                return sound.play(speed, cycle);
            }
            else
                return false;
        }

        public bool stopAnimation(string animationName)
        {
            AnimationSpecification anim = getAnimation(animationName);
            if (anim != null)
            {
                return anim.stop();
            }
            else
                return false;
        }

        public bool stopAllAnimations()
        {
            foreach (AnimationSpecification anim in getAnimations())
            {
                anim.stop();
            }
            return true;
        }

        public bool stopSound(string name)
        {
            SoundSpecification sound = getSound(name);
            if (sound != null)
            {
                return sound.stop();
            }
            else
                return false;
        }

        public void _addChildren(Entity valueEntity)
        {

            if (children.Contains(valueEntity))
                children.Add(valueEntity);
        }
        public void _removeChildren(Entity valueEntity)
        {
            if (children.Contains(valueEntity))
                children.Remove(valueEntity);

        }

        public PointSpecification getViewPoint(string name)
        {
            System.Console.WriteLine(this.name + " has " + viewPoints.Keys.Count + " viewpoints");

            if (viewPoints.ContainsKey(name))
                return viewPoints[name];

            foreach (Entity child in children)
            {
                PointSpecification vp = child.getViewPoint(name);
                if (vp != null)
                {
                    return vp;
                }
            }
            return null;
        }

        public void setParent(Entity e)
        {
            if (parent != null && parent.children.Contains(e))
            {
                parent.children.Remove(this);

            }
            //referentialPoint.Parent = null;
            parent = e;
            if (parent != null)
            {
                //	referentialPoint.Parent = parent.referentialPoint;

                if (parent.children.Contains(this))
                    parent.children.Add(this);
            }
        }

        public Entity Clone()
        {
            MascaretApplication.Instance.VRComponentFactory.Log(" ######## Cloning : " + this.name);

            Entity clone = new Entity(this.name, (EntityClass)(this.Classifier));
            
            /*foreach (Slot slot in this.Slots.Values)
            {

                //clone.Slots.Add(slot.getValue().getStringFromValue(), slot);

                //Slot s = new Slot();

                foreach (KeyValuePair<string, ValueSpecification> pair in slot.Values)
                {
                    //clone.Slots.Add(pair.Value.getStringFromValue(), slot);
                    ValueSpecification val = pair.Value.clone();
                    //s.addValue(val);
                }
                clone.Slots.Add(slot.getValue().getStringFromValue(), s);
            }*/

            foreach (KeyValuePair<string,Slot> slot in this.Slots)
            {
                
                MascaretApplication.Instance.VRComponentFactory.Log(" ######## Add Value for : " + slot.Value.DefiningProperty.name);
               
                //clone.addSlot(slot.Value.DefiningProperty);
                Slot s = clone.getProperty(slot.Value.DefiningProperty.name);
                
                foreach (KeyValuePair<string, ValueSpecification> pair in slot.Value.Values)
                {
                    ValueSpecification val = pair.Value.clone();
                    s.addValue(val);
                }
                
            }

            return clone;
        }

        
    }
}

